v1.5.x: build error due to warning in usnic provider

The travis CI build fails for v1.5.x branch with the following error:

```
  CC       prov/usnic/src/usnic_direct/src_libfabric_la-usd_device.lo
prov/usnic/src/usnic_direct/usd_dest.c:218:11: error: taking address of packed member 'uh_ip' of class or structure 'usd_udp_hdr' may result in an unaligned pointer value [-Werror,-Waddress-of-packed-member]
    ip = &dest->ds_dest.ds_udp.u_hdr.uh_ip;
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

e.g. https://travis-ci.org/ofiwg/libfabric/jobs/328299369 


@bturrubiates

--- a/prov/usnic/src/usdf_rdm.c
+++ b/prov/usnic/src/usdf_rdm.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2014-2016, Cisco Systems, Inc. All rights reserved.
+ * Copyright (c) 2014-2017, Cisco Systems, Inc. All rights reserved.
  *
  * This software is available to you under a choice of one of two
  * licenses.  You may choose to be licensed under the terms of the GNU
@@ -128,22 +128,25 @@ usdf_rdm_rdc_ready(struct usdf_rdm_connection *rdc, struct usdf_tx *tx)
 }
 
 static inline uint16_t
-usdf_rdm_rdc_hash_helper(uint16_t *ipaddr, uint16_t port)
+usdf_rdm_rdc_hash_helper(uint32_t ipaddr, uint16_t port)
 {
 	uint16_t hash_index;
 
-	hash_index = ipaddr[0];
-	hash_index ^= ipaddr[1];
+	uint16_t lower = (ipaddr & 0xFFFF);
+	uint16_t upper = (ipaddr >> 16);
+
+	hash_index = lower;
+	hash_index ^= upper;
 	hash_index ^= port;
 
 	return hash_index & USDF_RDM_HASH_MASK;
 }
 
+
 static inline uint16_t
 usdf_rdm_rdc_hash_hdr(struct usd_udp_hdr *hdr)
 {
-	return usdf_rdm_rdc_hash_helper((uint16_t *)&hdr->uh_ip.saddr,
-			hdr->uh_udp.source);
+	return usdf_rdm_rdc_hash_helper(hdr->uh_ip.saddr, hdr->uh_udp.source);
 }
 
 static inline int
@@ -154,18 +157,18 @@ usdf_rdm_rdc_hdr_match(struct usdf_rdm_connection *rdc, struct usd_udp_hdr *hdr)
 }
 
 static inline int
-usdf_rdm_rdc_addr_match(struct usdf_rdm_connection *rdc, uint16_t *ipaddr,
+usdf_rdm_rdc_addr_match(struct usdf_rdm_connection *rdc, uint32_t ipaddr,
 		 uint16_t port)
 {
-	return *(uint32_t *)ipaddr == rdc->dc_hdr.uh_ip.daddr &&
-	    port == rdc->dc_hdr.uh_udp.dest;
+	return ipaddr == rdc->dc_hdr.uh_ip.daddr &&
+	       port == rdc->dc_hdr.uh_udp.dest;
 }
 
 /*
  * Find a matching RDM connection on this domain
  */
 static inline struct usdf_rdm_connection *
-usdf_rdm_rdc_addr_lookup(struct usdf_domain *udp, uint16_t *ipaddr,
+usdf_rdm_rdc_addr_lookup(struct usdf_domain *udp, uint32_t ipaddr,
 		uint16_t port)
 {
 	uint16_t hash_index;
@@ -216,8 +219,7 @@ usdf_rdm_rdc_insert(struct usdf_domain *udp, struct usdf_rdm_connection *rdc)
 {
 	uint16_t hash_index;
 
-	hash_index = usdf_rdm_rdc_hash_helper(
-		(uint16_t *)&rdc->dc_hdr.uh_ip.daddr,
+	hash_index = usdf_rdm_rdc_hash_helper(rdc->dc_hdr.uh_ip.daddr,
 		rdc->dc_hdr.uh_udp.dest);
 	USDF_DBG_SYS(EP_DATA, "insert rdc %p at %u\n", rdc, hash_index);
 
@@ -231,8 +233,7 @@ usdf_rdm_rdc_remove(struct usdf_domain *udp, struct usdf_rdm_connection *rdc)
 	uint16_t hash_index;
 	struct usdf_rdm_connection *prev;
 
-	hash_index = usdf_rdm_rdc_hash_helper(
-		(uint16_t *)&rdc->dc_hdr.uh_ip.daddr,
+	hash_index = usdf_rdm_rdc_hash_helper(rdc->dc_hdr.uh_ip.daddr,
 		rdc->dc_hdr.uh_udp.dest);
 	USDF_DBG_SYS(EP_DATA, "remove rdc %p from %u\n", rdc, hash_index);
 
@@ -292,7 +293,7 @@ usdf_rdm_rdc_tx_get(struct usdf_dest *dest, struct usdf_ep *ep)
 
 	udp = tx->tx_domain;
 	rdc = usdf_rdm_rdc_addr_lookup(udp,
-		(uint16_t *)&dest->ds_dest.ds_dest.ds_udp.u_hdr.uh_ip.daddr,
+		dest->ds_dest.ds_dest.ds_udp.u_hdr.uh_ip.daddr,
 		dest->ds_dest.ds_dest.ds_udp.u_hdr.uh_udp.dest);
 
 	if (rdc == NULL) {

--- a/prov/usnic/src/usnic_direct/usd_dest.c
+++ b/prov/usnic/src/usnic_direct/usd_dest.c
@@ -207,26 +207,32 @@ usd_fill_udp_dest(
     uint32_t daddr_be,
     uint16_t dport_be)
 {
-    struct ether_header *eth;
-    struct iphdr *ip;
-    struct udphdr *udp;
+    struct ether_header eth = {
+        .ether_type = htons(0x0800)
+    };
 
-    eth = &dest->ds_dest.ds_udp.u_hdr.uh_eth;
-    memcpy(eth->ether_shost, dap->uda_mac_addr, ETH_ALEN);
-    eth->ether_type = htons(0x0800);
+    struct udphdr udp = {
+        .dest = dport_be
+    };
 
-    ip = &dest->ds_dest.ds_udp.u_hdr.uh_ip;
-    ip->ihl = 5;                       /* no options */
-    ip->version = 4;
-    ip->tos = 0;
-    ip->frag_off = 0;
-    ip->ttl = 8;
-    ip->protocol = IPPROTO_UDP;
-    ip->saddr = dap->uda_ipaddr_be;
-    ip->daddr = daddr_be;
+    struct iphdr ip = {
+        .saddr = dap->uda_ipaddr_be,
+        .daddr = daddr_be,
+        .protocol = IPPROTO_UDP,
+        .version = 4,
+        .frag_off = 0,
+        .ihl = 5, /* no options */
+        .tos = 0,
+        .ttl = 8
+    };
 
-    udp = &dest->ds_dest.ds_udp.u_hdr.uh_udp;
-    udp->dest = dport_be;
+    /* Workaround taking a pointer to an element of a packed structure due to
+     * warnings in Clang 4.0.1 and beyond.
+     */
+    memcpy(eth.ether_shost, dap->uda_mac_addr, ETH_ALEN);
+    dest->ds_dest.ds_udp.u_hdr.uh_eth = eth;
+    dest->ds_dest.ds_udp.u_hdr.uh_udp = udp;
+    dest->ds_dest.ds_udp.u_hdr.uh_ip = ip;
 }
 
 static int
@@ -353,14 +359,16 @@ usd_dest_set_udp_ports(
     struct usd_qp *src_uqp,
     uint16_t dest_port_be)
 {
-    struct usd_qp_impl *qp;
-    struct udphdr *udp;
+    struct usd_qp_impl *qp = to_qpi(src_uqp);
+    struct udphdr udp = {
+        .source = qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port,
+        .dest = dest_port_be
+    };
 
-    qp = to_qpi(src_uqp);
-    udp = &dest->ds_dest.ds_udp.u_hdr.uh_udp;
-    udp->source = qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
-    udp->dest = dest_port_be;
-
+    /* Workaround taking a pointer to an element of a packed structure due to
+     * warnings in Clang 4.0.1 and beyond.
+     */
+    dest->ds_dest.ds_udp.u_hdr.uh_udp = udp;
 }
 
 /*

